File: merge_sort.py
def merge_sort(arr):
    if len(arr) <= 1:
        return
    m = len(arr)//2
    left, right = arr[:m], arr[m:]
    merge_sort(left)
    merge_sort(right)
    merge(left, right, arr)


def merge(left, right, arr):
    i = l = r = 0
    while l < len(left) and r < len(right):
        if left[l] < right[r]:
            arr[i] = left[l]
            l += 1
            i += 1
        else:
            arr[i] = right[r]
            r += 1
            i += 1
    while l < len(left):
            arr[i] = left[l]
            l += 1
            i += 1
    while r < len(right):
            arr[i] = right[r]
            r += 1
            i += 1


x = [356, 97, 846, 215, 213, 301]
merge_sort(x)
print(x)

File: C++/merge_sort.cpp
#include <iostream>
#include <vector>

using namespace std;

void merge(vector<int>& left, vector<int>& right, vector<int>& arr){
    int i = 0, l = 0, r = 0;
    while (l < left.size() && r < right.size()){
        if (left[l] < right[r])
            arr[i++] = left[l++];
        else
            arr[i++] = right[r++];
    }
    while (l < left.size())
        arr[i++] = left[l++];
    while (r < right.size())
        arr[i++] = right[r++];
}

void merge_sort(vector<int>& arr, int length){
    int l = 0, m = (length+l)/2;
    if (length <= 1)
        return;
    vector<int> left, right;
    for (int i = 0; i < length; i++){
        if (i < m)
            left.push_back(arr[i]);
        else
            right.push_back(arr[i]);
    }
    merge_sort(left, left.size());
    merge_sort(right, right.size());
    merge(left, right, arr);
}

int main(){
    vector<int> x = {6,2,8,3,1,9};
    merge_sort(x, x.size());
    
    for (int i = 0; i < x.size(); i++){
        cout<<x[i]<<", ";
    }
    cout<<endl;
}